cron/src/main/scala/com/criteo/cuttle/cron/CronScheduler.scala

```diff
@@ -125,35 +124,54 @@ case class CronScheduler(logger: Logger) extends Scheduler[CronScheduling] {
   }
 
   private def run(job: CronJob, executor: Executor[CronScheduling]): IO[Completed] = {
-    def runAndRetry(job: Job[CronScheduling], scheduledAt: ScheduledAt, retryNum: Int, runNowUser: Option[Auth.User]): IO[Completed] =
+    def runNextPart(job: CronJob, scheduledAt: ScheduledAt, runNowUser: Option[Auth.User]): IO[Completed] =
       // don't run anything when job is paused
       if (state.isPaused(job)) {
         IO.pure(Completed)
       } else {
-        val runIO = for {
-          runInfo <- IO {
-            logger.debug(s"Sending job ${job.id} to executor")
-            val cronContext = CronContext(scheduledAt.instant)(retryNum)
-            executor.run(job, cronContext)
-          }
-          _ <- IO(runNowUser.fold(())(user => runInfo._1.streams.info(s"Run now request from ${user.userId}")))
-          _ <- IO(state.addNextExecutionToState(job, runInfo._1))
-          _ <- logState
-          completed <- IO.fromFuture(IO(runInfo._2))
-        } yield completed
-
-        runIO.recoverWith {
-          case e: Throwable =>
-            if (retryNum < job.scheduling.maxRetry) {
-              val nextRetry = retryNum + 1
-              logger.debug(s"Job ${job.id} has failed it's going to be retried. Retry number: $nextRetry")
-              runAndRetry(job, scheduledAt, nextRetry, runNowUser)
-            } else {
-              logger.debug(s"Job ${job.id} has reached the maximum number of retries")
-              IO.pure(Completed)
-            }
+        state.getNextParts(job).map{node: CronJobPart =>
+            for {
+               _ <- runAndRetry(job,
+                         node,
+                         scheduledAt,
+                         node.maxRetry,
+                         runNowUser)
+               _ <- IO(state.cronJobPartFinished(job, node))
+               result <- runNextPart(job, scheduledAt, runNowUser)
+           } yield result
+        }.toList.sequence_.map(_ => Completed)
```

Comment:
`sequence` here will run the IO sequentially. You can use `parSequence` instead since these steps can be parallelised: https://typelevel.org/cats-effect/datatypes/io.html#parsequence